ducktape/tests/action.py

```diff
@@ -0,0 +1,230 @@
+        self.result.append({'restype': restype, 'reason': reason, 'loc': loc})
+        self.logger.info('TestAction "%s" RESULT: %s' % (self.fullname, restype.upper()))
+
+    def passed (self, reason):
+        """ Pass this action with the provided reason """
+        self.add_result('pass', reason)
+        self.pass_cnt += 1
+        self.logger.info('Action "%s" PASSED: %s' % (self.fullname, reason))
+
+    def failed (self, reason):
+        """ Fail this action with the provided reason """
+        self.add_result('fail', reason)
+        self.fail_cnt += 1
+        self.logger.error('Action "%s" FAILED: %s' % (self.fullname, reason))
+
+    def warning (self, reason):
+        """ Logs a warning for the current action, adds the same as a 'WARN' result.
+            Use this for non-critical errors that may still affect the outcome of the test action,
+            e.g., environment anomalies. """
+        self.add_result('warn', reason)
+        self.warn_cnt += 1
+        self.logger.warning('Action "%s" WARNING: %s' % (self.fullname, warning))
+
+    def done (self):
+        """ Finishes the action with either passed or failed depending on if any
+            sub actions failed or not. """
+        if self.is_done():
+            return # Already done
+        failed = sum([x.fail_cnt for x in self.subs])
+        if failed > 0:
+            self.fail_cnt += 1
+        else:
+            self.pass_cnt += 1
+        self.duration = time.time() - self.start_time
+
+
+    def has_failed (self):
+        return self.fail_cnt > 0
+
+    def has_passed (self):
+        return self.pass_cnt > 0 and self.fail_cnt == 0
+
+    def is_done (self):
+        return self.has_failed() or self.has_passed()
+
+    def status (self):
+        """ Returns the status string for this action """
+        if not self.has_failed() and not self.has_passed():
+            return 'dnf'
+        elif self.has_failed():
+            return 'fail'
+        elif self.has_passed():
+            return 'pass'
+        else:
+            return 'partial'
+
+    def get_result (self):
+        """ Populate a result dict for this action and all its sub-actions.
+            Returns a list of such dicts. """
+        res = dict()
+        res['name'] = self.name
+        res['fullname'] = self.fullname
+        res['instance'] = self.instance
+        res['desc'] = self.desc
+        res['status'] = self.status()
+        res['results'] = self.result
+        res['warncnt'] = self.warn_cnt
+        res['start_time'] = self.start_time
+        res['duration'] = '%.3f' % float(self.duration)
+        res['subcnt'] = len(self.subs)
+
+        res_list = [res]
+        if len(self.subs) > 0:
+            for sub in self.subs:
+                res_list.extend(sub.get_result())
+        return res_list
+
+    def report (self, reporter):
+        """ *ActionReporter() feeder
+            FIXME: This should be consolidated with get_results() above """
```

Comment:
Agreed, I feel like the control is inverted here. This should just be a data object and the reporter should be responsible for traversing it's standard layout unless we think there will be multiple subtypes of `TestAction`.